File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
public:
vector<string>nums={"" , "" , "abc" , "def" , "ghi" , "jkl" , "mno" , "pqrs" , "tuv" , "wxyz"};

void help(string digits , vector<string>&ans , int i , string temp)
{
    if(i==digits.length())  
    {
        ans.push_back(temp);
        return ;
    }
    char number=digits[i];
    int n=number-'0';
    for(auto x : nums[n])
    {
        help(digits , ans , i+1 , temp+x);
    }
}
    vector<string> letterCombinations(string digits) {
vector<string>ans;
if(digits=="") return ans;
help(digits , ans , 0 , "");
return ans;
    }
};

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    List<String> mapping =  new ArrayList<>(List.of("","","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"));
    List<String> ans = new ArrayList<>();
    void genrateComination(String digits,String letters,int idx)
    {
        if(digits.length() == letters.length()){
            ans.add(letters);
            return;
        }
        String currDigit = mapping.get(digits.charAt(idx)-'0');
        for(int i=0;i<currDigit.length();++i)
            genrateComination(digits,letters+currDigit.charAt(i),idx+1);
        

    }
    public List<String> letterCombinations(String digits) {
        if (digits.length() == 0)
            return ans;
        genrateComination(digits,"",0);
        return ans;
    }
}